### PR TITLE
Avoid SIGABRT in CompAssembly::getComponentByName

### LINKED ISSUE #19002: LoadHelper should throw if component name does not exist (getComponentByName)
### Expected behavior

Mantid should not crash when an error occurs loading a file with incompatible IDF, incompatible in terms of unexpected component names.
`CompAssembly::getComponentByName` takes a string holding the component name as input. In case no component of that name exist, an exception should be thrown.

### Actual behavior

If a component of the given name does not exist in the IDF, `LoadHelper` tries to catch a runtime_error or a NotFoundError never thrown by `CompAssembly::getComponentByName`.

### Platforms affected

All

### PR DESCRIPTION
**Pull request by vreimund**
`CompAssembly::getComponentByName` takes a string holding the component name as input. In case no component of that name exist, an exception should be thrown rather than to stop Mantid executing (sigabrt).

**To test:**

Choose a file (likely a loader located in `DataHandling`), which uses `CompAssembly::getComponentByName` and its corresponding instrument definition file, then try loading.

For example:
- choose `LoadILLReflectrometry` and replace the component name  `uniq_detector` of its instrument definition file `D17_Definition.xml` with another name
- execute `LoadILLReflectometry`, it should result in an error but should not stop Mantid executing

Fixes #19002.

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Geometry/src/Instrument/CompAssembly.cpp [hunk 1]
```diff
@@ -344,7 +345,8 @@ CompAssembly::getComponentByName(const std::string &cname, int nlevels) const {
   } // while-end
 
   // If we have reached here then the search failed
-  return boost::shared_ptr<const IComponent>();
+  throw std::runtime_error("Error when trying to find the component " + cname +
+                           " , this component may not exist. \n");
 }
```
Comment (KarlPalmen) [inline Framework/Geometry/src/Instrument/CompAssembly.cpp]: One unit test requires invalid argument exception to be thrown if the component is not found (not runtime error).

Comment (KarlPalmen) [inline Framework/Geometry/src/Instrument/CompAssembly.cpp]: It is `SpecularReflectionCalculateThetaTest::test_throws_if_DetectorComponentName_unknown`.

Comment (reimundILL) [inline Framework/Geometry/src/Instrument/CompAssembly.cpp]: Thanks

**Comment by SimonHeybrock** [comment]
I am not sure that changing the behavior of `getComponentByName` is the right approach. I think there is client code that relies on the current behavior (returning a `nullptr`). Have you considered changing the code that causes the `SIGABRT` (is it attempting to dereference the `nullptr`?)?

**Comment by reimundILL** [comment]
I will change `LoadHelper.cpp` where the error occurs. I will leave the `getComponentByName` as it is. I was not sure which way to go, thank you for your opinion

**Comment by reimundILL** [comment]
Well, I did something wrong in git

**Pull request closed without merging**